src/lib/sticky-header/sticky-header-dir.ts

```diff
@@ -0,0 +1,381 @@
+    let unstuckCss: any = this.generateCssStyle(this.originalCss.zIndex,
+      'absolute', 'auto', '0', 'auto', '0', this.originalCss.width);
+    extendObject(this.element.style, unstuckCss);
+  }
+
+
+  /**
+   * 'sticker()' function contains the main logic of sticky-header. It decides when
+   * a header should be stick and when should it be unstuck. It will first get
+   * the offsetTop of the upper scrollable container. And then get the Start and End
+   * of the sticky-header's stickyRegion.
+   * The header will be stick if 'stickyRegion Start < container offsetTop < stickyRegion End'.
+   * And when 'stickyRegion End < container offsetTop', the header will be unstuck. It will be
+   * stick to the bottom of its stickyRegion container and being scrolled up with its stickyRegion
+   * container.
+   * When 'stickyRegion Start > container offsetTop', which means the header come back to the
+   * middle of the scrollable container, the header will be reset to its
+   * original CSS.
+   * A flag, isStuck. is used in this function. When a header is stick, isStuck = true.
+   * And when the 'isStuck' flag is TRUE, the sticky-header will not be repaint, which
+   * decreases the times on repainting sticky-header.
+   */
+  sticker(): void {
+    let currentPosition: number = this.upperScrollableContainer.offsetTop;
+
+    // unstuck when the element is scrolled out of the sticky region
+    if (this.isStuck &&
+      (currentPosition < this._containerStart || currentPosition > this._scrollFinish) ||
+      currentPosition >= this._scrollFinish) {
+      this.resetElement();
+      if (currentPosition >= this._scrollFinish) {
+        this.unstuckElement();
+      }
+      this.isStuck = false;    // stick when the element is within the sticky region
+    } else if ( this.isStuck === false &&
+      currentPosition > this._containerStart && currentPosition < this._scrollFinish) {
+      this.stickElement();
+    }
+  }
+
+  defineRestrictionsAndStick(): void {
+    this.defineRestrictions();
+    this.sticker();
+  }
+
+  /**
+   * This function is used to generate a variable which contains 7 css styles.
+   * @param zIndex
+   * @param position
+   * @param top
+   * @param right
+   * @param left
+   * @param bottom
+   * @param width
+   * @returns {{zIndex: any, position: any, top: any, right: any,
+   * left: any, bottom: any, width: any}}
+   */
+  generateCssStyle(zIndex:any, position:any, top:any, right:any,
+                   left:any, bottom:any, width:any): any {
+    let targetCSS = {
+      zIndex: zIndex,
+      position: position,
+      top: top,
+      right: right,
+      left: left,
+      bottom: bottom,
+      width: width,
+    };
+    return targetCSS;
+}
+
+
+  private getCssValue(element: any, property: string): any {
+    let result: any = '';
+    if (typeof window.getComputedStyle !== 'undefined') {
+      result = window.getComputedStyle(element, '').getPropertyValue(property);
```

Comment:
added 'padding' and 'stickyRegionHeight' variables to avoid calling 'getComputedStyle()' too many times (which is expensive).